### PR TITLE
8296954: G1: Enable parallel scanning for heap region remset

### PR DESCRIPTION
**Pull request by walulyai**
Hi all, 

Please review this change that allows parallel scanning of a heap region's remembered set.  More balanced work load distribution in cases where are cards are unevenly distributed among remembered sets.

Testing: Tier 1-3

Thanks

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296954](https://bugs.openjdk.org/browse/JDK-8296954): G1: Enable parallel scanning for heap region remset

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [b0891932](https://git.openjdk.org/jdk/pull/11173/files/b0891932f7e5f94090f4855e98fc66b0b148b134)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11173/head:pull/11173` \
`$ git checkout pull/11173`

Update a local copy of the PR: \
`$ git checkout pull/11173` \
`$ git pull https://git.openjdk.org/jdk pull/11173/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11173`

View PR using the GUI difftool: \
`$ git pr show -t 11173`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11173.diff">https://git.openjdk.org/jdk/pull/11173.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 1]
```diff
@@ -916,6 +916,9 @@ void G1RemSet::assert_scan_top_is_null(uint hrm_index) {
 void G1RemSet::prepare_region_for_scan(HeapRegion* r) {
   uint hrm_index = r->hrm_index();
 
+  // we call prepare_remset_for_scan unconditionally because optional evacuation doesn't not call this method again.
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:919]:
```suggestion
```

I do not think there is need for a comment here.

**Review by tschatzl — APPROVED** [review]
Lgtm.

File: src/hotspot/share/gc/g1/g1CardSet.cpp [hunk 2]
```diff
@@ -240,12 +241,15 @@ void G1CardSetCoarsenStats::print_on(outputStream* out) {
 
 class G1CardSetHashTable : public CHeapObj<mtGCCardSet> {
   using ContainerPtr = G1CardSet::ContainerPtr;
+  using CHTScanTask = CardSetHash::ScanTask;
 
+  const static uint BucketClaimSize = 16;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1CardSet.cpp:246]: Some comment on how `16` is derived would be nice and probably useful for future re-evaluation.

**Review by albertnetymk — APPROVED** [review]
I believe some performance results (benchmark setup, what metric is improved, etc) should be posted on the corresponding JBS ticket to better motivate this feature. The change looks fine.

**Comment by walulyai** [comment]
Thanks @tschatzl and @albertnetymk for the reviews.

/integrate

**Pull request closed without merging**